### DISCUSSION TITLE
How do you modify data in Prisma Migrate?

### DISCUSSION THREAD

**Discussion by hyunwoona** [prisma/prisma#10031] [Q&A] [answered]
How do you modify data in Prisma Migrate to fit the new schema without resetting the database?
For example, assume you had 'Joe Biden' for `full_name` field.
How do you change `full_name` to `first_name` (Joe) and `last_name` (Biden) based on the existing records?
Other ORMs let you write scripts to read from the existing columns and insert computed values to the new columns. I can't find the equivalent for Prisma. `seed` doesn't seem to be attached to a migration and resets the database, making it only suitable for seeding the initial data.

**Comment by austincrim** [comment]
Hey @hyunwoona!

Prisma Migrate does not have this functionality built-in, but you could write your own script to accomplish this. In your example, I would recommend using Prisma Migrate to add the `first_name` and `last_name` columns, run an update script to populate the new columns, then perform another migration that removes the `middle_name` column.

If you would like to see specific features added to Prisma Migrate, please fill out a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=). 

Thanks!

**Reply by hyunwoona** [reply]
Thanks, @austincrim.

Do you have any examples of such scripts? Ideally the script will be attached to each migration version so after running each migration, the script runs automatically.

I will file a feature request too since I think it's a very important feature.

**Reply by austincrim** [reply]
Unfortunately, I don't have any script examples on hand. 

However, my teammate pointed out perhaps a better approach for your use case. You could generate a migration file with `prisma migrate dev --create-only` and then write your migration script in the resulting SQL file. This way you can run the migration using Prisma Migrate like usual.

You might also be interested in [how RedwoodJS approaches data migrations](https://redwoodjs.com/docs/data-migrations).

In any case, thank you for the question and feature request, your feedback is super helpful!

**Reply by IsaiahByDayah** [reply]
@austincrim Not sure if you have much firsthand experience with RedwoodJS but do you know if it's possible to only use the data migration aspect of RedwoodJS? 

I'm pretty new to Prisma and like the _schema_ migration tool in Prisma a lot but now I'm wondering if there's a similar solution for _data_ migration that can easy work alongside Prisma + NextJS. Redwood's solution seemed to be just that but it looks like it's a part of their whole ecosystem and not something you could piece out. Would love to know what you recommend for devs using Prisma & NextJS!

One thing I thought about doing is just making an NextJS API route that can only be called if you have the right credentials and do the same logic that would be in Redwood's data migration. Its a bit funky and not really a "great" solution, but it would be effective in getting things done.

**Comment by sarink** [comment]
How can this feature still be missing 👀 

It seems like it would be trivial for prisma to at least add pre/post-migration hooks, something as simple as calling a `prisma/migrations/timestamp_name/before.js` before `migration.sql` and `prisma/migrations/timestamp_name/after.js` after `migration.sql`

**Reply by janpio** [reply]
Of course. But what would that solve? What code would this hook run for older migrations where the Prisma Client reflecting that database schema is not present any more? As usual, in the whole system this is a more complicated than it initially seems.

**Reply by sarink** [reply]
@janpio I agree, attempting to keep an archive of prisma clients around (or however you'd fix that complicated problem) doesn't seem feasible.

But we don't need prisma client and all its awesomeness to be used here. Just give developers an official hook to perform automated migrations in a language that isn't SQL. Something that doesn't force us to perform them manually on deploy, or write our own hacky workarounds for this very common problem (just look at some of the proposed solutions in this thread, or even the one taken by redwood). All of that is worse than the low-effort solution prisma could provide today.

It's possible that someone could attempt to import/require a potentially out-of-sync prisma client here, I think a simple line in the docs about the dangers of that would suffice. Doing this is still far better than the status quo.

**Reply by elliott-w** [reply]
> What code would this hook run for older migrations where the Prisma Client reflecting that database schema is not present any more?

@janpio I would always use the `$queryRaw` and `$executeRaw` methods on the prisma client.

Also I would like the ability to execute typescript files `before.ts` and `after.ts` as hooks too.

Hooks like these would be a seriously useful/helpful feature.

**Reply by sarink** [reply]
@janpio any update here?

**Reply by jharrell** [reply, marked as answer]
Hi @sarink, no updates currently. We have a [few open feature requests](https://github.com/prisma/prisma/issues?q=is:issue+is:open+label:kind/feature+label:%22topic:+data+migrations%22+) for issues similar to this. I'd recommend giving a 👍 to those that match the issues you are having! Thanks 😊

**Comment by magoz** [comment]
The docs recommend to run migrations for production at CI level.

I need to run a script between migrations:
- create a nullable column
- populate the data
- run another migration to make the column non nullable.

What's the recommended approach to achieve this?

Which solution did you end up going with @sarik, @hyunwoona ?

**Comment by revero-doug** [comment]
given there's already a shadow database required, what's the harm in allowing prisma client to be used during migration, then aggregate and spit out the SQL it generated into a sql file to be applied in prod? note that it's an experimental feature, and if it fails, tell the user it fails (and give a >0 exit code) and _then_ it's a userland problem. people will absolutely learn how to stay on the happy path if it means a more straightforward migration process 80+% of the time.

**Comment by HartS** [comment]
We're using Prisma since before migrations, and hacked in our own way to handle migrations with a continuous deployment script (which does other things also)

These only handle "rolling forward" though, so if we needed to rollback at some point we'd need to customize scripts for that

Basically, our CD scripts just check commits and perform setup as necessary. If there's a schema change in a commit *and* the hook symlinks have been modified in the same commit, it runs the before_push and after_push scripts around the prisma db push command.

It's a bit brittle, but lets us run migrations that actually use Prisma itself (in javascript) to performa necessary actions around the migration

**Reply by cawabunga** [reply]
I wrote a bash script that does checks and execute the hook:

```bash
#!/bin/bash

hook_file="db/post-hook.ts"

# List of our files to check
files=("db/schema.prisma" "$hook_file")

# Join our files to a string with | separator
files_joined=$(printf "|%s" "${files[@]}")
files_joined=${files_joined:1}

# Get the number of files to check
files_count=${#files[@]}

# Get a list of all changed files in the last commit
changed_files=$(git show --name-only --pretty="")

# Determine the number of changed files that match our list of files
files_num=$(printf "%s" "$changed_files" | grep -c -E "$files_joined")

# If all of our files were changed, execute the hook script
if [[ "$files_num" -eq "$files_count" && -f "$hook_file" ]]; then
    yarn tsx "$hook_file"
fi
```

Unfortunately such approach doesn't work if we decide to accumulate some changes (for example a few PRs) before deploying, as it checks only the last commit.

**Comment by revero-doug** [comment]
I'm seeing a major design limitation (I'd argue flaw) behind the unwillingness to support any of the suggested approaches to supporting data migrations: prisma doesn't internally support joins or subqueries and instead only works using N round trips.

This yields some notable constraints, including but not limited to:
1. generating SQL data migrations is impractical
    a. must have access to a database with not only the right schema, but records with production IDs
2. running prisma client code during migrations will be non-deterministic in terms of SQL commands yielded

Thoughts:
- constraint 1 is a hard constraint that I don't see changing until or unless prisma supports an option to create single SQL queries from single prisma client calls. every time this has been requested or discussed, there's been pretty hardcore pushback, which I think is a mistake.
- constraint 2 is not really a serious issue. if that non-determinism works in the application (which it does, or prisma would be unusable), it should work in migrations.

If we're agreed constraint 2 is a unimportant, just expose migration hooks and make the users happy (pretty please).

**Comment by xiaolongkipsi** [comment]
I would like to have a hook too. We are searching a way to write data migration in javascript instead of raw SQL

**Comment by bistacos** [comment]
+1 to all this, performing a data migration in a language other than SQL seems like a primary concern for a migration tool. The `create new nullable field --> populate using custom logic --> make field non-nullable` flow at minimum is super common. Some kind of before/after hook or the ability to run custom migration script <programming-language-x> is necessary for us to be able to use Prisma Migrate in production.

I think the priority should be higher here: it doesn't make a lot of sense for a tool that considers SQL to be tedious and worth abstracting to ask you to default to super-complicated SQL if you need to make a data migration.

Thoughts @janpio ? Is the main hurdle here that exposing a compatible Prisma Client in the middle of a set of migrations is really complex to implement? I'd be happy with just being able to execute raw DB queries and do some data munging/normalization in javascript/python/whatever, I don't necessarily need Prisma Client. Yes it introduces complications between up/down but I've never worked in an environment where we didn't need custom data migrations, unfortunately.

**Reply by janpio** [reply]
We realize that. But we are to busy with other things right now to figure out the details and even make a statement what the hurdle might be. Right now it is that we just do not have the capacity and are working on other priorities.

**Reply by bistacos** [reply]
Well not the news I was hoping for but I appreciate you keeping us up to date on priorities and likely next features.

**Reply by sarink** [reply]
Well, there's still a reasonable low-effort solution that could be a huge improvement https://github.com/prisma/prisma/discussions/10031#discussioncomment-3753052

**Comment by jbgrunewald** [comment]
I'm just going to throw my opinion into the ring and say a way to write data migrations that are applied alongside model changes (in TS or JS) would be great. Coming from Type ORM, I definitely miss being able to write these sort of migrations to have them code reviewed and applied in our regular CI processes. 

I do acknowledge that we could maintain a separate process to apply data migrations in the CI pipeline after merging from a code review. However, I'm quite sure we'd end up duplicating much of the same details tracked by Prisma when model migrations are applied.

**Reply by HartS** [reply]
@jbgrunewald does Type ORM allow you to write migrations in TS/JS? I thought I took a look at that and saw it only allows you to do SQL

**Reply by HartS** [reply]
Just took a look; indeed, it looks like typeORM generates the migrations as a series of SQL queries driven by javascript. Presumably you can also insert additional code there, though I'm not clear on whether it all gets invoked in the same transaction (and consequently, fails to apply atomically if any failure occurs during the migration)

If it *does* do this, it's a pretty big deal IMO

**Reply by jbgrunewald** [reply]
@HartS I know for sure they support JS. I don't recall if TS was supported. You can definitely write your own logic in the hook functions they support. Normally you would update your entity classes, then generate a migration from those changes. However, you can write a custom migration where you also read and write data to do data migrations.

**Comment by HartS** [comment]
Looked into RedwoodJS a bit today, which uses Prisma under the hood.

However, they do add some [additional tooling](https://redwoodjs.com/docs/data-migrations) around Prisma's migrations that make add JS/TS data migrations into the mix, in a sensible way.

I wonder if this can be extracted into a separate package that can be used in conjunction with Prisma

**Comment by kockar96** [comment]
Maybe someone will find this article helpful:
https://medium.com/@bryantbrock/data-migrations-with-prisma-94909cb0c0c0

**Comment by redareda9** [comment]
Just came from Rails and super surprising to see that Prisma doesn't have such a basic function. 
ActiveRecord was as simple as having an up and down function in the migration. 
No need to hack it by building scripts manually and adding them to the CI

**Comment by sarink** [comment]
Prisma team: This is too complicated! You're not thinking it through!
Every other ORM: Provides simple before/after hooks that users are happy with

**Comment by Softjey** [comment]
Hi everyone! 👋

I’ve developed a small library to help address this problem: [Prisma Data Migrations](https://www.npmjs.com/package/prisma-data-migrations?activeTab=readme). This library enhances Prisma’s migration system by enabling the use of post-migration scripts for data migrations alongside schema changes.

You can find the repository here: [GitHub Repository](https://github.com/Softjey/prisma-data-migrations).

Feel free to open issues, report bugs, or suggest improvements. I’m also open to pull requests if you’d like to contribute to the project!

Looking forward to your feedback! 🚀

**Reply by huksley** [reply]
:+1:  this is interesting but quite heavyweight, IMO. can it just parse configs out of schema.prisma etc and support post.sql post.sh post.ts/js in th migration folder, that would be simple enough.

**Reply by Softjey** [reply]
You can use just post scripts without generating from prisma.schema. It's a feature that allows you to use typed Prisma ORM for migrations, but it's not mandatory.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
